cpp/include/hf/flexpolyline.h

```diff
@@ -4,302 +4,536 @@
  * SPDX-License-Identifier: MIT
  * License-Filename: LICENSE
  */
-#include <string>
-#include <cmath>
 #include <array>
+#include <cmath>
+#include <optional>
 #include <stdexcept>
+#include <string>
 #include <tuple>
-
+#include <variant>
+#include <vector>
+
+/// # Flexible Polyline encoding
+///
+/// The flexible polyline encoding is a lossy compressed representation of a list of coordinate
+/// pairs or coordinate triples. It achieves that by:
+///
+/// 1. Reducing the decimal digits of each value.
+/// 2. Encoding only the offset from the previous point.
+/// 3. Using variable length for each coordinate delta.
+/// 4. Using 64 URL-safe characters to display the result.
+///
+/// The encoding is a variant of [Encoded Polyline Algorithm Format]. The advantage of this encoding
+/// over the original are the following:
+///
+/// * Output string is composed by only URL-safe characters, i.e. may be used without URL encoding
+///   as query parameters.
+/// * Floating point precision is configurable: This allows to represent coordinates with precision
+///   up to microns (5 decimal places allow meter precision only).
+/// * It allows to encode a 3rd dimension with a given precision, which may be a level, altitude,
+///   elevation or some other custom value.
+///
+/// ## Specification
+///
+/// See [Specification].
+///
+/// [Encoded Polyline Algorithm Format]:
+/// https://developers.google.com/maps/documentation/utilities/polylinealgorithm
+///
+/// [Specification]: https://github.com/heremaps/flexible-polyline#specifications
+///
+/// ## Example
+///
+/// ```cpp
+/// #include< hf/flexpoly.h>
+///
+/// // encode
+/// std::vector<std::tuple<double, double>> coordinates = {{50.1022829, 8.6982122},
+///                                                        {50.1020076, 8.6956695},
+///                                                        {50.1006313, 8.6914960},
+///                                                        {50.0987800, 8.6875156}};
+/// hf::Polyline polyline = hf::Polyline2d{std::move(coordinates), hf::Precision::DIGITS5};
+///
+/// std::string encoded;
+/// auto error = hf::polyline_encode(polyline, encoded));
+/// assert(!error);
+/// assert(encoded == "BFoz5xJ67i1B1B7PzIhaxL7Y");
+///
+/// // decode
+/// hf::Polyline result;
+/// auto decode_error = hf::polyline_decode(encoded, result);
+/// assert(!decode_error);
+/// ```
 namespace hf {
 
-const uint8_t FORMAT_VERSION=1;
+/// Coordinate precision in the polyline
+///
+/// Represents how many digits are to be encoded after the decimal point, e.g.
+/// precision 3 would encode 4.456787 as 4.457.
+///
+/// Supported values: `[0,16)`
+enum class Precision : uint8_t {
```

Comment:
I'm all in favor of type safety and keeping hardcoded constants out of the cod, but I feel this approach is a little over the top and needlessly complex and repetitive. I'd just use `uint8_t`s and check if they are <= 15. Or maybe make `Precision` not an `enum` but a class with a `uint8_t` member and have its c'tor check whether the argument is <= 15?

Reply:
Better?

Reply:
Hm, I was thinking more along the lines of (sorry, my C++ skills are a bit rusty):
```
class Precision {
public: 
    Precision(uint32_t p) {
        if (p > 15) throw std::invalid_argument("...");
        m_precision = static_cast<uint8_t>(p);
    }
private:
    uint8_t  m_precision;
}
```
and dumping the entire definition of constants and `precision_from_u32()` which is just very repetitive and IMHO hurts readability more than it helps.

Reply:
The current version using the factory pattern since it is often best to avoid exception in C++

Reply:
Oh, I seem to have missed the change... yes, I actually like this version :)